### PR TITLE
improve gateway symlink handling

### LINKED ISSUE #6203: internalWebError: operation not supported
#### Version information:

```yaml
go-ipfs version: 0.4.20-rc1-4fd81e373-dirty
Repo version: 7
System version: amd64/linux
Golang version: go1.12.2
```

#### Description:

When I try to access [this hash](http://127.0.0.1:8080/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD) specifically, I get `internalWebError: operation not supported`. It also happens with plenty other public nodes.

The hash was found here : https://github.com/ipfs/package-managers/issues/18#issuecomment-475227762

#### Affected

* https://gateway.ipfs.io/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://gateway.pinata.cloud/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://gateway.serph.network/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://ipfs.eternum.io/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://ipfs.greyh.at/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://ipfs.infura.io/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://ipfs.io/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://ipfs.sloppyta.co/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://ipfs.wa.hle.rs/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD

#### Ok

* https://cloudflare-ipfs.com/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://ipfs.deo.moe/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD
* https://ipfs.jes.xxx/ipfs/QmThJ4k554iT3B7SZonmVMwspHALiHGThhqaem7nij1wQD

### PR DESCRIPTION
**Pull request by Stebalien**
* Correctly calculate symlink sizes.
* Fix listing of directories with symlinks.
* Handle files without known sizes.

Note: This _doesn't_ provide actual symlink support. The gateway treats symlinks
like text files.

fixes #6203

### CODE REVIEW

**Comment by Stebalien** [comment]
@senden9, @NatoBoram

File: core/corehttp/lazyseek.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+	size       int64
+	offset     int64
+	realOffset int64
+}
+
+func (s *lazySeeker) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case io.SeekEnd:
+		return s.Seek(s.size+offset, io.SeekStart)
```
Comment (hsanjuan) [inline core/corehttp/lazyseek.go]: isn't it `s.size-offset` ?

Comment (Stebalien) [reply core/corehttp/lazyseek.go]: If we're seeking from the end, the offset will already be negative.

**Comment by senden9** [comment]
Build and tested with commit 361bb5fb5e34837a40a6feddf4a1da751e981388.
Full "symlink support" would be the best but I understand that you want hack it in right now. But I can see now the directory in the gateway web UI without error messages!

File: core/corehttp/lazyseek.go (hunk 1 above)
Comment (Stebalien) [reply core/corehttp/lazyseek.go]: Actually, this _was_ wrong. It should have been `s.size-1-offset`.

File: core/corehttp/lazyseek.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	// If we're past the end, EOF.
+	if s.offset >= s.size {
+		return 0, io.EOF
+	}
+
+	// actually seek
+	for s.offset != s.realOffset {
+		off, err := s.reader.Seek(s.offset, io.SeekStart)
+		if err != nil {
+			return 9, err
```
Comment (dirkmc) [inline core/corehttp/lazyseek.go]: Is returning `9` intentional?

Comment (Stebalien) [reply core/corehttp/lazyseek.go]: Definitely not.

**Comment by Stebalien** [comment]
Thanks for catching that (and bugging me to write a test)!

**Comment by Stebalien** [comment]
(ok, there's definitely a bug here, I'll look into it when I get a chance)

**Pull request merged**